uint/src/uint.rs

```diff
@@ -512,6 +553,15 @@ macro_rules! construct_uint {
 				return true;
 			}
 
+			// Whether this fits u64.
+			#[inline]
+			fn fits_word(&self) -> bool {
```

Comment:
This looks like the reverse of https://github.com/paritytech/parity-common/blob/master/uint/src/uint.rs#L215 – maybe we can save some code there? And I wonder if the `unroll!` is still warranted, and if the comment about inlining is still relevant on modern rustc.

Reply:
> maybe we can save some code there? 

There is also a difference in that in it starts from 1nd index, not 0th, so I'm not sure how are you suggesting to reuse it. Initially I wrote `self.bits() <= 64`, but there was a small (~2%), but consistently measurable difference.

Reply:
I made a change to `unroll!` to work with non-zero indices [here](https://github.com/Vurich/crunchy/pull/4), I'm curious if it makes any difference? :)

Reply:
`unroll` v0.2.2 is out so I checked if it makes a difference (went back to commit 0b3844e1a) and it looks like unrolling is quite a significant win for `U256`:

```
     Running /Users/dvd/dev/parity/parity-common/target/release/deps/bigint-abf18d123360ed8c
u256_mul//(115792089237316195423570985008687907853269984665640564039457584007913129639935, 1)
                        time:   [3.3480 ns 3.3644 ns 3.3843 ns]
                        change: [+36.959% +41.024% +44.814%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
u256_mul//(3, 18446744073709551615)
                        time:   [3.3805 ns 3.3993 ns 3.4202 ns]
                        change: [+38.603% +43.937% +49.933%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe
u256_mul//(21674844646682989462120101885968193938394323990565507610662749, 173)
                        time:   [3.4109 ns 3.4509 ns 3.4963 ns]
                        change: [+34.791% +38.484% +41.754%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe

u512_mul//(1, 0)     time:   [8.5398 ns 8.5914 ns 8.6427 ns]
                        change: [+12.422% +16.167% +20.151%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) high severe
u512_mul//(18446744073709551615, 4294967296)
                        time:   [8.4936 ns 8.5363 ns 8.5772 ns]
                        change: [+9.2618% +13.348% +17.802%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe
u512_mul//(3759751734479964094783137206182536765532905409829204647089173492, 25097755455336819385...
                        time:   [24.957 ns 25.029 ns 25.111 ns]
                        change: [-4.7843% -1.5953% +1.9514%] (p = 0.40 > 0.05)
                        No change in performance detected.
Found 17 outliers among 100 measurements (17.00%)
  2 (2.00%) low mild
  9 (9.00%) high mild
  6 (6.00%) high severe
u512_mul//(452312848583266388373324160190187140051835877600158453279131187530910662655, 452312848...
                        time:   [25.200 ns 25.326 ns 25.461 ns]
                        change: [-7.0673% -4.0808% -1.6396%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe
```

Reply:
Thanks David, will add a commit with `unroll` 0.2.2.